File: SPOJ/EaglesAndDogs.java
import java.util.*;

public class EaglesAndDogs {

    private static class Edge {
        int src;
        int dst;
        int wt;

        Edge(int src, int dst, int wt) {
            this.src = src;
            this.dst = dst;
            this.wt = wt;
        }
    }
    public static void main(String[] args) {
        Scanner scn = new Scanner(System.in);
        int t = scn.nextInt();
        
        while(t-- > 0) {
            int n = scn.nextInt();

            // BUILDING GRAPH
            ArrayList<Edge>[] graph = new ArrayList[n];
            for( int idx = 0 ; idx < n ; idx++ )
                graph[idx] = new ArrayList<>();

            for(int idx = 0 ; idx < n - 1 ; idx++) {
                int vtx1 = scn.nextInt() - 1; 
                int vtx2 = scn.nextInt() - 1;
                int wt = scn.nextInt();

                graph[vtx1].add(new Edge(vtx1, vtx2, wt));
                graph[vtx2].add(new Edge(vtx2, vtx1, wt));
            }
            
            // APPLYING DFS TO FIND MAX WEIGHT PATH
            int[] dp = new int[n];
            Arrays.fill(dp, -1);   // DOGS AND EAGLES NEVER BE -1
            for( int vtx = 0 ; vtx < n ; vtx++ ) { 
                int vis[] = new int[n]; 
                Arrays.fill(vis, -1);
                dp[vtx] = dfs(graph, vtx, vis);
            }
        
            // DISPLAYING OUTPUT
            for(int idx = 0 ; idx < n ; idx++) 
                System.out.print(dp[idx] + " ");
            System.out.println();
        }
    }

    private static int dfs(ArrayList<Edge>[] graph, int cvtx, int[] vis) {
        vis[cvtx] = 0;
        for(Edge edge : graph[cvtx]) 
            if(vis[edge.dst] == -1)
                vis[cvtx] = Math.max(vis[cvtx], dfs(graph, edge.dst, vis) + edge.wt);
        
        return vis[cvtx];
    }
}

File: SPOJ/EaglesAndDogs.cpp
#include <iostream>
#include <vector>
#include <array>

using namespace std;


int
dfs(vector<pair<int,int> > graph[], int cvtx, int parent) {
    int maxlength = 0;
    for(int idx = 0 ; idx < graph[cvtx].size() ; idx++) {
        pair<int, int> p = graph[cvtx][idx];
        if(p.first != parent) {
            int cpath = dfs(graph, p.first, cvtx) + p.second;
            maxlength = maxlength > cpath ? maxlength : cpath;
        }
    }

    return maxlength;
}

int
main(int arc, char* argv[]) {
    int t;
    cin >> t;

    while(t--) {
        int n;
        cin >> n;
        vector<pair<int, int> > graph[n];

        // CREATING GRAPH FROM USER INPUT
        for(int iter = 0 ; iter < n - 1; iter ++) {
            int vtx1, vtx2, wt;
            cin >> vtx1 >> vtx2 >> wt;

            vtx1 -- ;
            vtx2 --;
            
            graph[vtx1].push_back(make_pair(vtx2, wt));
            graph[vtx2].push_back(make_pair(vtx1, wt));
        }

        // APPLYING DFS AT EVERY POINT
        for(int vtx = 0 ; vtx < n ; vtx++) {
            cout << dfs(graph, vtx, -1) << " ";
        }
        cout << endl;

    }

    return 0;
}